### PR TITLE
[2.7] bpo-29960 _random.Random corrupted on exception in setstate(). …

### PR DESCRIPTION
**Pull request by Mariatta**
…(GH-1019).

(cherry picked from commit 9616a82e7802241a4b74cf7ae38d43c37bf66e48)

### CODE REVIEW

**Comment by Mariatta** [comment]
@serhiy-storchaka Can you please check if I backported this correctly? There was a conflict in _randommodule.c which I have to resolve manually. Just making sure if everything is still in order.
Thanks :)

File: Modules/_randommodule.c [hunk 1]
```diff
@@ -341,6 +341,7 @@ random_setstate(RandomObject *self, PyObject *state)
     int i;
     unsigned long element;
     long index;
+    uint32_t new_state[N];
```
Comment (serhiy-storchaka) [inline Modules/_randommodule.c]: `uint32_t` is C99. Use `unsigned long` instead.

File: Modules/_randommodule.c [hunk 2]
```diff
@@ -357,7 +358,7 @@ random_setstate(RandomObject *self, PyObject *state)
         element = PyLong_AsUnsignedLong(PyTuple_GET_ITEM(state, i));
         if (element == (unsigned long)-1 && PyErr_Occurred())
             return NULL;
-        self->state[i] = element & 0xffffffffUL; /* Make sure we get sane state */
+        new_state[i] = (uint32_t)element;
```
Comment (serhiy-storchaka) [inline Modules/_randommodule.c]: Leave the 2.7 code at the right hand of assignment.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]

**Comment by serhiy-storchaka** [comment]
Why you have removed the `cherry-pick for 2.7` label @brettcannon?

**Comment by brettcannon** [comment]
@serhiy-storchaka there's an email to python-committers about it, but the label was removed from GitHub entirely so I didn't specifically remove it  just for this PR.

File: Modules/_randommodule.c (hunk 1 above)
Comment (Mariatta) [reply Modules/_randommodule.c]: Thanks, changed to `unsigned long`.

File: Modules/_randommodule.c (hunk 2 above)
Comment (Mariatta) [reply Modules/_randommodule.c]: Changed.

**Review by serhiy-storchaka — APPROVED** [review]

**Pull request merged**